# AUTOGENERATED FILE - DO NOT MODIFY!
# This file generated by Djinni from example.djinni

from djinni.support import MultiSet # default imported in all files
from djinni.exception import CPyException # default imported in all files
from djinni.pycffi_marshal import CPyObject, CPyRecord, CPyString

from dh__list_string import ListStringHelper
from PyCFFIlib_cffi import ffi, lib

from djinni import exception # this forces run of __init__.py which gives cpp option to call back into py to create exception

from item_list import ItemList

class ItemListHelper:
    @staticmethod
    def release(c_ptr):
        assert c_ptr in c_data_set
        c_data_set.remove(ffi.cast("void*", c_ptr))

    @ffi.callback("struct DjinniObjectHandle *(struct DjinniRecordHandle *)")
    def get_item_list_f1(cself):
        try:
            _ret = CPyObject.fromPy(ListStringHelper.c_data_set, CPyRecord.toPy(None, cself).items)
            assert _ret != ffi.NULL
            return _ret
        except Exception as _djinni_py_e:
            CPyException.setExceptionFromPy(_djinni_py_e)
            return ffi.NULL

    @ffi.callback("struct DjinniRecordHandle *(struct DjinniObjectHandle *)")
    def python_create_item_list(items):
        py_rec = ItemList(
            CPyObject.toPy(ListStringHelper.c_data_set, items))
        return CPyRecord.fromPy(ItemList.c_data_set, py_rec) #to do: can be optional?

    @ffi.callback("void (struct DjinniRecordHandle *)")
    def __delete(dh):
        assert dh in ItemList.c_data_set
        ItemList.c_data_set.remove(dh)

    @staticmethod
    def _add_callbacks():
        lib.item_list_add_callback_get_item_list_f1(ItemListHelper.get_item_list_f1)
        lib.item_list_add_callback_python_create_item_list(ItemListHelper.python_create_item_list)
        lib.item_list_add_callback___delete(ItemListHelper.__delete)

ItemListHelper._add_callbacks()

